Implement lock screen widgets UI

tzef opened the pull request:
This is the second PR for adding the stats widget to the lock screen
1.  https://github.com/wordpress-mobile/WordPress-iOS/pull/20309 (✅ Approved)
2. https://github.com/wordpress-mobile/WordPress-iOS/pull/20312 (In reviewing) 👈 you're here!
- UI layout for **site-selected** status (DONE)
- Add `minimumScaleFactor`(DONE)
- Add `PreviewProvider` for `LockScreenSingleStatView` (DONE)
3. https://github.com/wordpress-mobile/WordPress-iOS/pull/20342 (TBD)
- Add localization key for "views today" (WIP)
- Add dynamic type support and change layout if needed (WIP)
- Raise and track the discussion about localization concerns (TODO)
4. https://github.com/wordpress-mobile/WordPress-iOS/pull/20353 (✅ Approved)
5. https://github.com/wordpress-mobile/WordPress-iOS/pull/20371 (✅ Approved)
6. https://github.com/wordpress-mobile/WordPress-iOS/pull/20317 (✅ Approved)
7. https://github.com/wordpress-mobile/WordPress-iOS/pull/20368 (✅ Approved)
8. https://github.com/wordpress-mobile/WordPress-iOS/pull/20399 (✅ Approved)
9. https://github.com/wordpress-mobile/WordPress-iOS/pull/20405 (✅ Approved)
(Confirm the data display on UI correctly in this phase)
10. https://github.com/wordpress-mobile/WordPress-iOS/pull/20422 (✅ Approved)
11. https://github.com/wordpress-mobile/WordPress-iOS/pull/20427 (In Reviewing)

## Description

1. Implemented the widget view based on the design specification. Other statuses apart from the "site-selected" will be handled in the following pull requests.
2. Added a view provider to decouple the `LockScreenStatsWidgetsView` from the content logic. Other widgets in the future could reuse the same view provider for displaying different data, such as `comments` and `likes` in the same layout.
3. Extract `ViewModel` and `ViewModelMapper` and added to the WordPress target for unit testing because the app extension is not able to run unit tests for now.
4. Extract `statsUrl` for home widget data from private to the extension to share with lock screen widgets, also add the extension to WordPress target for testable.

## Images & Videos
| Widget in lock screen | Widget in selection page|
| - | - |
|![WidgetInLockScreen](https://user-images.githubusercontent.com/3096210/224727980-cc175040-6e78-4e22-a3ca-84a0c3cc9f66.png)|![WidgetInSelection](https://user-images.githubusercontent.com/3096210/224727998-6b125174-02cc-445a-9d32-532e73af9bd5.png)|

## Testing instructions
For the issue about widget not reflecting the default website data, it might relate to cache or refresh notification in the data store, which I'm not pretty sure about for now, but I think can defer the bug to the PR about handling the refresh mechanism, let this PR focus on UI rendering.

- When the user has not logged in or without the default website:
the widget on the selection page should display placeholder data for the user's reference.

- When the user logs in and has a default website:
the widget on the selection page should display the stats data from the default website.

- When the user adds a widget and had more than one website:
should be able to switch to another website in the widget configuration and display the corresponding stats data.

- When the user clicks the widget: 
should redirect to  today stats page for the widget selected website

## Regression Notes
1. Potential unintended areas of impact
Adding lock screen widget flow

2. What I did to test those areas of impact (or what existing automated tests I relied on)
The testing instructions in PR https://github.com/wordpress-mobile/WordPress-iOS/pull/20309

3. What automated tests I added (or what prevented me from doing so)
`WidgetsViewModelMapperTests` to test that the view model was correctly mapped from home widget data.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

staskus commented on WordPress/WordPressStatsWidgets/LockScreenWidgets/LockScreenStatsWidget.swift:
```diff
@@ -0,0 +1,36 @@
+
+    var body: some WidgetConfiguration {
+        IntentConfiguration(
+            kind: AppConfiguration.Widget.Stats.lockScreenTodayViewsKind,
```
What are your thoughts regarding `AppConfiguration.Widget.Stats.lockScreenTodayViewsKind`? 

It is used to reload specific widgets once the data loads. Is your plan to have only one flag like this, multiple flags like this or making it dynamic?

tzef replied:
My idea is to create multiple flags which are a 1-1 relationship with every widget style, not only following the existing convention, I also think it's safer to reduce the risk about miss deleting of the registered widgets, the dynamic kind might unconsciously remove some widgets to let some widgets suddenly disappear from the user's screen.

staskus replied:
Okay, got it.

staskus commented on WordPress/WordPressStatsWidgets/LockScreenWidgets/Models/LockScreenWidgetViewModelMapper.swift:
```diff
@@ -0,0 +1,37 @@
+        )
+    }
+
+    func getStatesURL() -> URL? {
```
```suggestion
    func getStatsURL() -> URL? {
```

tzef replied:
Oh, sorry for the typo :(
Thanks for the catching.

staskus commented on WordPress/WordPressStatsWidgets/LockScreenWidgets/Models/LockScreenWidgetViewModelMapper.swift:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    func getStatesURL() -> URL? {
+        if let todayData = data as? HomeWidgetTodayData {
```
`getViews()` works for both `HomeWidgetTodayData` and `HomeWidgetAllTimeData`.  However `getStateURL` only checks for `HomeWidgetTodayData`. Is there a reason for this?

tzef replied:
At first, I only consider the existing use case.
I think I should add for other types of widgets as well. 
thanks for the reminder

staskus replied:
Okay, since we're building the feature step by step it's good to define the scope as clearly as possible in the description. Since getStatsURL and getViews differed I didn't know what the intention was.

staskus commented on WordPress/WordPressStatsWidgets/LockScreenWidgets/Models/LockScreenWidgetViewModelMapper.swift:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    func getViews() -> String {
+        if let todayData = data as? HomeWidgetTodayData {
```
If we're going to do this switching for both `url` and `views`, what do you think about having a `protocol` that is implemented buy `HomeWidgetTodayData` and `HomeWidgetAllTimeData`? Lets say: 

```swift
protocol LockScreenStatsWidgetData {
    var statsURL: URL { get }
    var views: Int { get }
    <...>
}
```

and then this mapper depending on `LockScreenStatsWidgetData` rather than `HomeWidgetData`. Would that work? 🤔

tzef replied:
I considered to have `LockScreenStatsWidgetData`, the reason I didn't do it in this PR is that would need to create a standalone `TimeLineProvider`.

As we discussed, I will add the Provider and Entry because some logic is not necessary for the lock screen, so I think it's good to create `LockScreenStatsWidgetData` as your suggestions, but I'd like to defer it to the PR for creating lock screen Provider and Entry, let this PR focus on UI layer.

staskus replied:
Okay, thanks for explaining. I see you mentioned it in the TODO so we can leave it for another PR.

staskus commented on WordPress/WordPressStatsWidgets/LockScreenWidgets/Views/LockScreenSingleStatView.swift:
```diff
@@ -0,0 +1,34 @@
+import WidgetKit
+
+@available(iOS 16.0, *)
+struct LockScreenSingleStatView: View {
```
Would it be possible to add `PreviewProvider` at the bottom? It would be a nice way to see how the view reacts to different types of inputs.

tzef replied:
Good suggestion!! I forgot it, thanks for the reminder.

staskus commented on WordPress/WordPressStatsWidgets/LockScreenWidgets/Views/LockScreenSingleStatView.swift:
```diff
@@ -0,0 +1,34 @@
+                        .frame(maxWidth: .infinity, alignment: .leading)
+                        .font(.system(size: 20, weight: .bold))
+                        .foregroundColor(.white)
+                    Text("\(viewModel.title) \(viewModel.dateRange)")
```
I wonder about different languages. Whether `title + dateRange` always makes sense, or there'll be languages where `dateRange + title` or any other combination makes sense.

Maybe we should just have a single String here which then can be appropriately localized? WDYT?

tzef replied:
Great point, I never think about this concern.
I prefer to treat it as "two" fields to make the feature simple for now because if we treat it as "one" field, then we need to provide many translations for the combination (field + date range)

And can probably improve the design later because one of the opinions from designers is there's another better design for them. Just pick this one as the safest design for the first iteration.
 
for example: 
| Large number | date range on top |
| - | - |
|<img width="123" alt="image" src="https://user-images.githubusercontent.com/3096210/225338653-975fbd02-094f-4bea-b5e7-1cd333c53e6d.png">|<img width="232" alt="image" src="https://user-images.githubusercontent.com/3096210/225339427-ccff8207-20e5-47c5-bb55-e343f8e6d4ac.png">|

So I suggest noting this concern as the discussion topic in the next phase, not blocking the current development progress, what do you think?

staskus replied:
> Great point, I never think about this concern.
I prefer to treat it as "two" fields to make the feature simple for now because if we treat it as "one" field, then we need to provide many translations for the combination (field + date range)

Got it, I understand where you're coming from. Since I12n is an important topic, I would like to take it into consideration. If we translate these labels separately and put one next to each other, the formulations may look incorrect in other languages. From the user's perspective, they won't see 2 different labels, they will see 2 words next to each other.

> And can probably improve the design later because one of the opinions from designers is there's another better design for them. Just pick this one as the safest design for the first iteration.

Yes, definitely we are always open to improvements 👍. However, since we chose a particular style of design for the moment, we should make sure it works well in different languages! 

> So I suggest noting this concern as the discussion topic in the next phase, not blocking the current development progress, what do you think?

I'm okay leaving it outside the scope of this PR but I prefer resolving the localization concern in this phase.

tzef replied:
No problem! I think it's good to create a specific key for this design style, and have the discussion in parallel, I will track the progress, so we can update the implementation depending on the new conclusion.

staskus commented on WordPress/WordPressStatsWidgets/LockScreenWidgets/Views/LockScreenSingleStatView.swift:
```diff
@@ -0,0 +1,34 @@
+                    Text(viewModel.value)
+                        .frame(maxWidth: .infinity, alignment: .leading)
+                        .font(.system(size: 20, weight: .bold))
+                        .foregroundColor(.white)
```
Maybe we could introduce `minimumScaleFactor` here? Although a number is unlikely to be so large it's cheap to avoid such corner-cases.

tzef replied:
Nice suggestion, sorry for I forgetting it, this point was also raised by the designer as well.
I temporarily choose 0.8 for 11 font size and 0.5 for 20 font size, I think if too smaller might become illegible.
feel free to give a better number of suggestions.

staskus commented on WordPress/WordPressTest/Widgets/WidgetsViewModelMapperTests.swift:
```diff
@@ -0,0 +1,47 @@
+import XCTest
+@testable import WordPress
+
+final class WidgetsViewModelMapperTests: XCTestCase {
```
👍

staskus commented on WordPress/WordPressStatsWidgets/LockScreenWidgets/Views/LockScreenStatsWidgetsView.swift:
```diff
@@ -1,7 +1,7 @@
 import WidgetKit
 
-protocol LockScreenStatsWidgetsDataSource {
+protocol LockScreenStatsWidgetsViewProvider {
```
Thanks for a change 👍 Reads clearer to me.

staskus reviewed (commented):
👍 Thank you for the PR! I'm leaving a couple of comments to discuss.

🔨 Since we're reviewing the code as well as testing the solution, we're used to adding `Testing instructions` to Pull Requests. The `Acceptance Criteria` is also great but it doesn't always explain "how" to make a change or reach a certain change, to the instructions can be helpful! Also, the recordings could be provided showing how it's achieved.

✅ **When the user has not logged in or without the default website, the widget on the selection page should display placeholder data for the user's reference.**

 ⚠️ **When the user logs in and has a default website, the widget on the selection page should display the stats data from the default website.**
 
 - The selection page shows me default site correctly, however after selecting the widget, the widget appears with "Build later" until I select a site from the configuration.
 
 ✅ **When the user adds a widget and had more than one website, should be able to switch to another website in the widget configuration and display the corresponding stats data.**
 - I agree that it's not intuitive how-to configuration is reached. Although it's out of scope.
 
 ✅ **When the user clicks the widget, should redirect to today stats page for the widget-selected website**

staskus commented on WordPress/WordPressStatsWidgets/LockScreenWidgets/Views/LockScreenStatsWidgetsView.swift:
```diff
@@ -0,0 +1,28 @@
+                .buildSiteSelectedView(data)
+                .widgetURL(viewProvider.statsURL(data))
+        default:
+            Text("Build Later")
```
Are there further improvements coming for this "Build later" label in other PRs?

tzef replied:
Yes, other status will be implemented in the following PRs, I added TODO comments to avoid confusion.

tzef committed:
Extract statsUrl from private to extension for reusable with lock screen widget

tzef committed:
Implement single stat view for today views widget

Add view provider for handling building view for site selected status
Add ViewModelMapper for converting viewModel from home widget data

tzef committed:
Update LockScreenStatsWidgetsView body to get the view from view provider

tzef committed:
Add mapper test to verify the viewModel converting from widget data correctly

Add mapper and viewModel to WordPress target because app extension is not able to run unitTest

tzef commented:
Thank you for the solid suggestions, and sorry for any mistakes due to not being careful enough.

I have updated the format of the testing instructions to follow the convention.

Regarding the issue of data not coming from the default website, it works on my simulator, but I believe it exists because I have not yet handled the data refresh part. Therefore, I suggest we limit the testing to the UI perspective. I will put all related test cases in the PR about handling data refresh.

Additionally, I suggest we also move the language concern to the end of the process. It would be beneficial to have input from the designer so I can raise a design revision post to see if anything could be improved based on the first version.

For other issues, I have already fixed and reorganized the commits. Please let me know if I missed anything. Thank you.

tzef committed:
Fix getStatsURL typo

staskus commented:
> Regarding the issue of data not coming from the default website, it works on my simulator, but I believe it exists because I have not yet handled the data refresh part. Therefore, I suggest we limit the testing to the UI perspective. I will put all related test cases in the PR about handling data refresh.

@tzef  I see that many questions come from the fact that I'm not sure if it's intended to be implemented now or in a later PR. It would be helpful to track a list of the done, ongoing, and planned tasks for these rectangular stats widgets so I would understand the thinking. As an example, you could create a PR for a parent branch where you would list all the tasks. This way you wouldn't need to list these task dependencies in the every PR you make. How does that sound?

staskus commented:
> Additionally, I suggest we also move the language concern to the end of the process. It would be beneficial to have input from the designer so I can raise a design revision post to see if anything could be improved based on the first version.

i12n is an important topic so we should definitely think about it when building UIs. Clarifying with the designers is a good idea but we shouldn't be afraid of making decisions ourselves if we see a problem! 👍

staskus commented on WordPress/WordPressStatsWidgets/LockScreenWidgets/Views/LockScreenSingleStatView.swift:
```diff
@@ -0,0 +1,57 @@
+                VStack(alignment: .leading) {
+                    Text(viewModel.siteName)
+                        .frame(maxWidth: .infinity, alignment: .leading)
+                        .font(.system(size: 11))
```
🔤 If possible, we should try supporting [dynamic type](https://github.com/wordpress-mobile/WordPress-iOS/blob/trunk/docs/dynamic-type-guidelines.md) so users who have the text size increased or decreased in the accessibility settings would have reflected that in our apps in our widgets. I think setting the fixed system size does not support dynamic type out of the box. Could you check if dynamic type support is possible for these lock screen widgets?

[You can check more of our guidelines in the docs](https://github.com/wordpress-mobile/WordPress-iOS/tree/trunk/docs).

tzef replied:
I admit I intended to use fixed font because the space is not enough in the lock screen widget, and sorry for I didn't highlight this decision during the design discussion.

I just implicitly said the font size of the number is smaller than the original design due to space issues. 
That's why I promoted another design to get the ability to enlarge the font.
<img width="272" alt="image" src="https://user-images.githubusercontent.com/3096210/225554861-67c20ee9-557c-40b8-8588-4a5502762f5f.png">

For now, we don't have enough space to enlarge the font (only remain the edge padding spaces).
so my suggested solution is how about we change the layout to horizontal for the large font? (the right side one in the screenshot)

staskus replied:
Thanks for the explanation!

I'm checking my current lock widgets, and they are not reacting to dynamic type changes either. As you explained, there's very little space to deal with. Even if we move to a horizontal layout, we will get other types of issues.

Let's stick to the layout we've chosen. 👍

tzef commented:
>As an example, you could create a PR for a parent branch where you would list all the tasks. This way you wouldn't need to list these task dependencies in the every PR you make. How does that sound?

That's a great idea. Especially we did step-by-step implementation for the feature, I can't create PR for the parent branch feature/19411-add-lock-screen-stats-widgets due to I'm not the owner, so I will update the roadmap to every PRs and highlight the progress.

>i12n is an important topic so we should definitely think about it when building UIs. Clarifying with the designers is a good idea but we shouldn't be afraid of making decisions ourselves if we see a problem! 👍

Understand, then for this case, I think we can create a new key for "views today" to make sure this style works well, for other fields with different date ranges can postpone after the new widget is decided.

And I'm not familiar with the new translation task flow here, I guess we also have a remote service (such as [webtranslateit](https://webtranslateit.com/), [lokalise](https://lokalise.com/)) and automatically sync the key and value for each release process?  I will try to find the related document, if you can provide the link that'd be helpful, thanks 🙏

In the meanwhile, I will raise this concern to Jetpack to ask for the designer's input, does it good for you?

staskus reviewed (commented):
Thanks for the changes and for adding TODOs for the tasks unrelated to UI! 👍

I would still want to see if we could improve the accessibility of this widget in the scope of the UI. Left the comment next to the code.

staskus commented:
> That's a great idea. Especially we did step-by-step implementation for the feature, I can't create PR for the parent branch feature/19411-add-lock-screen-stats-widgets due to I'm not the owner, so I will update the roadmap to every PRs and highlight the progress.

Thanks!

> And I'm not familiar with the new translation task flow here, I guess we also have a remote service (such as [webtranslateit](https://webtranslateit.com/), [lokalise](https://lokalise.com/)) and automatically sync the key and value for each release process? I will try to find the related document, if you can provide the link that'd be helpful, thanks 🙏

These are the localization guidelines in our repository docs: https://github.com/wordpress-mobile/WordPress-iOS/blob/trunk/docs/localization.md. You don't need to worry about the remote localization services, everything is done during the release process. 

> In the meanwhile, I will raise this concern to Jetpack to ask for the designer's input, does it good for you?

Yes, it works.

staskus reviewed (approved):
Okay, thanks for all the answers and including a list of planned tasks on the top of PR.

We can move on to the other PRs and other work.

staskus commented:
@tzef I merged the PR.
